From e928621a0bbd010b75624c77105ce31f2b8d2faf Mon Sep 17 00:00:00 2001
Message-ID: <e928621a0bbd010b75624c77105ce31f2b8d2faf.1779348625.git.lorenzo@kernel.org>
In-Reply-To: <e15783f7c987e199ecf80b3d858ed5a86d33c508.1779348625.git.lorenzo@kernel.org>
References: <e15783f7c987e199ecf80b3d858ed5a86d33c508.1779348625.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Sat, 13 Dec 2025 09:45:09 +0100
Subject: [PATCH 12/13] net: airoha: Add ethtool priv_flags callbacks

Introduce ethtool priv_flags callbacks in order to allow the user to
select if the configured device will be used as hw lan or wan and enable
or disable gdm2 loopback (used for hw QoS).

Tested-by: Madhur Agrawal <madhur.agrawal@airoha.com>
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c  | 173 ++++++++++++++++++++++
 drivers/net/ethernet/airoha/airoha_regs.h |   1 +
 2 files changed, 174 insertions(+)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -2157,10 +2157,7 @@ static void airoha_update_netdev_feature
 			struct airoha_gdm_dev *iter_dev = port->devs[j];
 			struct net_device *netdev;
 
-			if (!iter_dev || iter_dev == dev)
-				continue;
-			if (rcu_access_pointer(iter_dev->qdma) !=
-			    rcu_access_pointer(dev->qdma))
+			if (!iter_dev)
 				continue;
 
 			netdev = netdev_from_priv(iter_dev);
@@ -2172,6 +2169,16 @@ static void airoha_update_netdev_feature
 	}
 }
 
+static void airoha_qdma_user_get(struct airoha_qdma *qdma)
+{
+	qdma->users++;
+}
+
+static void airoha_qdma_user_put(struct airoha_qdma *qdma)
+{
+	qdma->users--;
+}
+
 static int airoha_dev_open(struct net_device *netdev)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
@@ -2215,7 +2222,7 @@ static int airoha_dev_open(struct net_de
 
 	airoha_dev_set_xmit_frame_size(netdev);
 	port->users++;
-	qdma->users++;
+	airoha_qdma_user_get(qdma);
 
 	if (!airoha_is_lan_gdm_dev(dev) &&
 	    airoha_ppe_is_enabled(eth, 1))
@@ -2248,7 +2255,7 @@ static int airoha_dev_stop(struct net_de
 					    FE_PSE_PORT_DROP);
 
 	qdma = airoha_qdma_deref(dev);
-	qdma->users--;
+	airoha_qdma_user_put(qdma);
 	airoha_update_netdev_features(dev);
 
 	phylink_stop(dev->phylink);
@@ -2409,17 +2416,28 @@ static void airoha_dev_set_qdma(struct a
 	struct net_device *netdev = netdev_from_priv(dev);
 	struct airoha_qdma *cur_qdma, *qdma;
 	struct airoha_eth *eth = dev->eth;
+	bool running = netif_running(netdev);
 	int i, ppe_id;
 
 	/* QDMA0 is used for lan ports while QDMA1 is used for WAN ports */
 	qdma = &eth->qdma[!airoha_is_lan_gdm_dev(dev)];
 	cur_qdma = airoha_qdma_deref(dev);
 
-	if (cur_qdma)
+	if (cur_qdma) {
 		netif_tx_stop_all_queues(netdev);
+		if (running && cur_qdma != qdma) {
+			int id = cur_qdma - &eth->qdma[0];
+
+			if (netdev->features & NETIF_F_GRO_HW)
+				airoha_fe_lro_disable(eth, id);
+			airoha_qdma_user_put(cur_qdma);
+		}
+	}
 
 	rcu_assign_pointer(dev->qdma, qdma);
 	netdev->irq = qdma->irq_banks[0].irq;
+	if (running && cur_qdma != qdma)
+		airoha_qdma_user_get(qdma);
 	synchronize_rcu();
 
 	ppe_id = !airoha_is_lan_gdm_dev(dev) && airoha_ppe_is_enabled(eth, 1);
@@ -2438,6 +2456,12 @@ static void airoha_dev_set_qdma(struct a
 
 	if (cur_qdma)
 		netif_tx_wake_all_queues(netdev);
+
+	if (cur_qdma != qdma) {
+		airoha_update_netdev_features(dev);
+		if (running && (netdev->features & NETIF_F_GRO_HW))
+			airoha_dev_lro_enable(dev);
+	}
 }
 
 static int airoha_dev_init(struct net_device *netdev)
